### PR TITLE
Build + Support Py39 and Various CI Updates

### PR DESCRIPTION
**Pull request by JacksonMaxfield**
I was originally trying to get `numcodecs` for py39 + macOS building [as I have a build failing](https://github.com/AllenCellModeling/aicsimageio/pull/185/checks?check_run_id=1755420390), but, in working on this:

* I saw https://github.com/zarr-developers/numcodecs/pull/265 (Updating CI / moving from Travis to GitHub Actions)
* I also saw https://github.com/zarr-developers/numcodecs/issues/269 (blosc upgrade)
  * But then saw in zarr dev community call notes that it may be preferred to move from blosc submodule to blosc pip install.
  * And all the tests + wheel builds on MacOS + Py39 so not sure where to fall on this

In addressing the CI changes, I tried to figure out the `travis` and `appveyor` prior setups but it was a bit tricky. The one missing thing is `coveralls` which has a [GitHub Action](https://github.com/marketplace/actions/coveralls-github-action) that we could use but didn't know how you wanted to handle that. (Sidenote: other option is to move to codecov which is _very_ easy to setup in my experince)

Reference Issues:
https://github.com/zarr-developers/numcodecs/issues/262
https://github.com/zarr-developers/numcodecs/issues/261
https://github.com/zarr-developers/numcodecs/issues/269

TODO:
* [x] Unit tests and/or doctests in docstrings
* [x] ``tox -e py39`` passes locally
* [x] Docstrings and API docs for any new/modified user-facing classes and functions
* [x] Changes documented in docs/release.rst
* [x] ``tox -e docs`` passes locally
* [ ] Test coverage to 100% (Coveralls passes)

### CODE REVIEW

**Comment by joshmoore** [comment]
Oh wow. Nice, @JacksonMaxfield. Briefly on the trade-off: @jakirkham suggests that moving to the upstream wheels will be doable and advantageous but will require a number of changes throughout the build system. Once done, it would mean that this repo could be python only. (But no one has yet started on that work) But it looks like you are close so worth moving forward?

If so, moving to codecov is completely valid and has been done elsewhere in the org. For that matter, I could see ignoring leaving it temporarily as a TODO as we figure out the other questions.

**Comment by JacksonMaxfield** [comment]
> But it looks like you are close so worth moving forward?

Sorry, is this saying I should add the blosc wheel support to this PR or saying handling it in the future. I am a bit afraid to touch any of the `c` / `pyx` files not going to lie 😅

File: setup.py [hunk 1]
```diff
@@ -322,18 +322,7 @@ def run_setup(with_extensions):
...
-            'Intended Audience :: Science/Research',
-            'License :: OSI Approved :: MIT License',
-            'Programming Language :: Python',
-            'Topic :: Software Development :: Libraries :: Python Modules',
-            'Operating System :: Unix',
-            'Programming Language :: Python :: 3',
-            'Programming Language :: Python :: 3.6',
-            'Programming Language :: Python :: 3.7',
-            'Programming Language :: Python :: 3.8',
+            "Programming Language :: Python :: 3.9",
```
Comment (Carreau) [inline setup.py:325]: did you meant to delete all of those ?

File: .github/workflows/wheel.yaml [hunk 2]
```diff
@@ -9,7 +9,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-18.04, windows-latest, macos-latest]
+        os: [ubuntu-18.04, ubuntu-20.04, windows-latest, macos-latest]
```
Comment (Carreau) [inline .github/workflows/wheel.yaml:12]: Is it really necessary, or can we build wheel only on one of the two linux ?

**Comment by Carreau** [comment]
I've disable making Appveyor mandatory if we move to gh-action so this can be merged even with appveyor failing. I'm +1 on moving to codecov as well.

**Comment by Carreau** [comment]
> Sorry, is this saying I should add the blosc wheel support to this PR or saying handling it in the future. I am a bit afraid to touch any of the `c` / `pyx` files not going to lie 😅

I think we can wait for the migration to the blosc wheel, i will likely not need to touch c/pyx beyond deleting them though, but might need minimal API change internally; I think there was a discussion to at least bump the minor version when we do that.

File: setup.py (hunk 1 above)
Comment (JacksonMaxfield) [reply setup.py:325]: Whoops. Definitely didn't mean to delete these.

File: .github/workflows/wheel.yaml (hunk 2 above)
Comment (JacksonMaxfield) [reply .github/workflows/wheel.yaml:12]: I don't think both are needed. So I will just replace with `ubuntu-latest`.

File: setup.py (hunk 1 above)
Comment (JacksonMaxfield) [reply setup.py:337]: Ahhhh my editor changed `'` to `"`, hope thats okay.

**Comment by joshmoore** [comment]
> Sorry, is this saying I should add the blosc wheel support to this PR or saying handling it in the future

The latter.

> I think we can wait for the migration to the blosc wheel,

Exactly.

;) Sorry for being unclear.

**Comment by jakirkham** [comment]
This looks really good! Thanks for working on this Jackson 😄

Did we want to add Windows to GitHub Actions as well?

**Comment by JacksonMaxfield** [comment]
> Did we want to add Windows to GitHub Actions as well?

That should be pretty easy to do. Just didn't see any windows CI stuff in the repo besides the wheel building. I will add it in later today.

**Comment by JacksonMaxfield** [comment]
@jakirkham I spoke too soon. It seems that there is semi-known weird numpy + pickle interaction which results in a single one of the tests failing. Specifically the backwards compatibility test on pickles.

[StackOverflow](https://stackoverflow.com/questions/58194852/modulenotfounderror-no-module-named-numpy-core-multiarray-r) explaining the issue. I think it has to do with the `"\r\n"` / windows carrage return character.

Alright if I skip that test on Windows?

**Comment by jakirkham** [comment]
Thanks Jackson! 😄

Yeah that sounds like a good idea. Maybe we can mark it as `xfail` on Windows? Could you please raise an issue as well so we can track fixing that in the future? 🙂

**Comment by JacksonMaxfield** [comment]
> Yeah that sounds like a good idea. Maybe we can mark it as `xfail` on Windows? Could you please raise an issue as well so we can track fixing that in the future? slightly_smiling_face

Sounds good! Should be good to go w/ the latest Windows CI passing

File: numcodecs/tests/test_pickles.py [hunk 3]
```diff
@@ -43,7 +43,13 @@ def test_repr():
...
 
-@pytest.mark.skipif(sys.byteorder != 'little',
-                    reason='Pickle does not restore byte orders')
+# Details on xfail
+# https://stackoverflow.com/questions/58194852/modulenotfounderror-no-module-named-numpy-core-multiarray-r
+@pytest.mark.skipif(
+    sys.byteorder != "little", reason="Pickle does not restore byte orders"
+)
+@pytest.mark.xfail(
+    sys.platform == "win32", reason="Pickle fails to read w/ Windows carriage return"
```
Comment (jakirkham) [inline numcodecs/tests/test_pickles.py:52]: Would suggest linking the issue like so

```suggestion
    sys.platform == "win32", reason=(
        "Pickle fails to read w/ Windows carriage return "
        "( https://github.com/zarr-developers/numcodecs/issues/271 )"
    )
```

**Review by jakirkham — COMMENTED** [review]
Looks good. Thanks for adding that 🙂

Made a small suggestion to the xfail comment. Otherwise LGTM

**Comment by jakirkham** [comment]
I've also checked the box to disable AppVeyor builds when `appveyor.yml` is missing. So hopefully that status check will disappear.

**Comment by jakirkham** [comment]
Thanks again Jackson! 😄

**Comment by JacksonMaxfield** [comment]
Thank you all for making this package to begin with :slightly_smiling_face:

**Comment by JacksonMaxfield** [comment]
Hey @jakirkham any chance to bumpversion and release to support py39?

**Comment by joshmoore** [comment]
Tagged & pushed. 🤞🏼

**Comment by JacksonMaxfield** [comment]
Last update before I head to bed: my tests now pass w/ Mac Py39 support so hopefully not jinxing it when I say "seems like everything worked" :slightly_smiling_face:

**Comment by jakirkham** [comment]
Should now have wheels and Conda packages.

**Comment by joshmoore** [comment]
@jakirkham : on that note, it looked like in docs/contributing.rst that I should take the checksum of the local .tar.gz?? Can you _explicame_ a bit?

**Comment by jakirkham** [comment]
Ah that's outdated. We can drop that. I think that was when we were still using GitHub archives instead of PyPI as the source of conda-forge packages

**Comment by jakirkham** [comment]
Fixing the doc in PR ( https://github.com/zarr-developers/numcodecs/pull/272 )

**Pull request merged**